src/nvim/tui/input.c

```diff
@@ -282,27 +289,56 @@ static bool handle_focus_event(TermInput *input)
 
 static bool handle_bracketed_paste(TermInput *input)
 {
-  if (rbuffer_size(input->read_stream.buffer) > 5 &&
-      (!rbuffer_cmp(input->read_stream.buffer, "\x1b[200~", 6) ||
-       !rbuffer_cmp(input->read_stream.buffer, "\x1b[201~", 6))) {
-    bool enable = *rbuffer_get(input->read_stream.buffer, 4) == '0';
-    // Advance past the sequence
-    rbuffer_consumed(input->read_stream.buffer, 6);
-    if (input->paste_enabled == enable) {
-      return true;
+  RBuffer *rbuf = input->read_stream.buffer;
+  if (rbuffer_size(rbuf) > 5
+      && (!rbuffer_cmp(rbuf, "\x1b[200~", 6)
+          || !rbuffer_cmp(rbuf, "\x1b[201~", 6))) {
+    bool enable = *rbuffer_get(rbuf, 4) == '0';
+
+    if (enable && input->paste_started) {
+      // Received a bogus "paste start" after paste was already started.
+      return false;
+    }
+
+    rbuffer_consumed(rbuf, 6);  // Advance past the sequence
+    input->paste_started = enable;
+    if (enable) {
+      loop_schedule(&loop, event_create(1, apply_pastepre, 0));
+    } else {
+      enqueue_input(input, PASTEPOST_KEY, sizeof(PASTEPOST_KEY) - 1);
```

Comment:
> Do we currently have any better mechanism for placing an event at a precise location in the input stream (I'm guessing the answer is "no", hence why we still have these pseudokeys)

No we don't.

> I am wondering if it would be a good idea to provide some internal generic mechanism for placing K_EVENT in the input stream and associating it with a callback.

K_EVENT is not meant to be used by UIs, it is only a general mechanism for waking nvim when an asynchronous event happens. An UI placing a bunch of keys in the input stream is an asynchronous event, but not "certain key gets processed"(And the only use case I could think for this is toggling paste, so I'm not sure it would be useful)